chainstate/tx-verifier/src/transaction_verifier/input_output_policy/constraints_accumulator.rs

```diff
@@ -13,705 +13,584 @@
-            OutputValue::Coin(Amount::from_atoms(staked_atoms - fee_atoms)),
-            Destination::AnyoneCanSpend,
-            OutputTimeLock::ForBlockCount(required_maturity_distance.to_int() as u64),
-        )];
-
-        let mut constraints_accumulator = ConstrainedValueAccumulator::new();
-
-        constraints_accumulator
-            .process_inputs(
-                &chain_config,
-                BlockHeight::new(1),
-                pledge_getter,
-                &inputs,
-                &input_utxos,
-            )
-            .unwrap();
-
-        constraints_accumulator.process_outputs(&chain_config, &outputs).unwrap();
-
-        assert_eq!(
-            constraints_accumulator.consume().unwrap().into_atoms(),
-            fee_atoms
-        );
+        Ok(())
     }
 
-    // Check that it's allowed to pay fees from spending a delegation share
-    // by providing smaller total outputs than inputs
-    #[rstest]
-    #[trace]
-    #[case(Seed::from_entropy())]
-    fn allow_fees_from_spend_share(#[case] seed: Seed) {
-        let mut rng = make_seedable_rng(seed);
-
-        let chain_config = common::chain::config::Builder::new(ChainType::Mainnet)
-            .consensus_upgrades(NetUpgrades::regtest_with_pos())
-            .build();
-        let required_maturity_distance =
-            chain_config.staking_pool_spend_maturity_distance(BlockHeight::new(1));
-
-        let delegation_id = DelegationId::new(H256::zero());
-        let delegated_atoms = rng.gen_range(100..1000);
-        let fee_atoms = rng.gen_range(1..100);
-
-        let pledge_getter = |_| Ok(None);
-
-        let inputs_utxos = vec![None];
-        let inputs = vec![TxInput::from_account(
-            AccountNonce::new(0),
-            AccountSpending::DelegationBalance(delegation_id, Amount::from_atoms(delegated_atoms)),
-        )];
-
-        let outputs = vec![TxOutput::LockThenTransfer(
-            OutputValue::Coin(Amount::from_atoms(delegated_atoms - fee_atoms)),
-            Destination::AnyoneCanSpend,
-            OutputTimeLock::ForBlockCount(required_maturity_distance.to_int() as u64),
-        )];
-
-        let mut constraints_accumulator = ConstrainedValueAccumulator::new();
-
-        constraints_accumulator
-            .process_inputs(
-                &chain_config,
-                BlockHeight::new(1),
-                pledge_getter,
-                &inputs,
-                &inputs_utxos,
-            )
-            .unwrap();
-
-        constraints_accumulator.process_outputs(&chain_config, &outputs).unwrap();
-
-        assert_eq!(
-            constraints_accumulator.consume().unwrap().into_atoms(),
-            fee_atoms
-        );
+    pub fn combine(&mut self, other: ConstrainedValueAccumulator) -> Result<(), IOPolicyError> {
+        merge_amount_maps(&mut self.unconstrained_value, other.unconstrained_value)?;
+        merge_amount_maps(&mut self.timelock_constrained, other.timelock_constrained)?;
```

Comment:
Similar comment as for `merge_amounts_map`. If you could take `self` by value and return it, it would make for a more functional interface.